quarkchain/cluster/master.py

```diff
@@ -107,119 +108,127 @@ def __init__(self, header, peer):
         """raise on any error so that sync() will close peer connection"""
         if self.__has_block_hash(self.header.get_hash()):
             return
 
-        # descending height
-        block_header_chain = [self.header]
+        ancestor = await self.__find_ancestor()
+        if ancestor is None:
+            raise RuntimeError("Cannot find common ancestor with max fork lengh {}".format(
+                self.root_state.root_config.MAX_STALE_ROOT_BLOCK_HEIGHT_DIFF)
+            )
 
-        while not self.__has_block_hash(block_header_chain[-1].hash_prev_block):
-            block_hash = block_header_chain[-1].hash_prev_block
-            height = block_header_chain[-1].height - 1
+        while self.header.height > ancestor.height:
+            limit = min(
+                self.header.height - ancestor.height,
+                self.root_state.root_config.MAX_STALE_ROOT_BLOCK_HEIGHT_DIFF
+            )
+            resp = await self.__download_block_header_and_check(ancestor.height + 1, 0, limit)
 
-            # abort if we have to download super old blocks
-            if self.root_state.tip.height - height > self.max_staleness:
-                Logger.warning(
-                    "[R] abort syncing due to forking at super old block {} << {}".format(
-                        height, self.root_state.tip.height
-                    )
-                )
+            block_header_chain = resp.block_header_list
+            if len(block_header_chain) == 0:
+                Logger.info("Remote chain reorg causing empty root block headers")
                 return
 
-            download_start_time = time_ms()
-            Logger.info(
-                "[R] downloading block header list from height {} with hash {}".format(
-                    height, block_hash.hex()
-                )
-            )
-            block_header_list = await asyncio.wait_for(
-                self.__download_block_headers(block_hash), SYNC_TIMEOUT
-            )
-            Logger.info(
-                "[R] downloaded block header list from height {} with hash {}, use {} ms".format(
-                    height, block_hash.hex(), time_ms() - download_start_time
-                )
-            )
-            self.__validate_block_headers(block_header_list)
-            for header in block_header_list:
-                if self.__has_block_hash(header.get_hash()):
-                    break
-                block_header_chain.append(header)
+            # Remote root block is reorg with new tip and new height (which may be lower than that of current)
+            # Setup end as the new height
+            if resp.root_tip != self.header:
+                self.header = resp.root_tip
 
-        block_header_chain.reverse()
+            if block_header_chain[0].hash_prev_block != ancestor.get_hash():
+                # TODO: Remote chain may reorg, may retry the sync
+                raise RuntimeError("Bad peer sending incorrect canonical headers")
```

Comment:
if this could be caused by legitimate reorg, maybe drop the "Bad peer" part

Reply:
Done.